pandas/core/groupby/groupby.py

```diff
@@ -847,43 +857,69 @@ def _transform_should_cast(self, func_nm):
 
     def _cython_transform(self, how, numeric_only=True, **kwargs):
         output = collections.OrderedDict()
-        for name, obj in self._iterate_slices():
+        names = []  # type: List[Hashable]
+        for idx, (name, obj) in enumerate(self._iterate_slices()):
             is_numeric = is_numeric_dtype(obj.dtype)
             if numeric_only and not is_numeric:
                 continue
 
             try:
-                result, names = self.grouper.transform(obj.values, how, **kwargs)
+                result, _ = self.grouper.transform(obj.values, how, **kwargs)
             except NotImplementedError:
                 continue
+
             if self._transform_should_cast(how):
-                output[name] = self._try_cast(result, obj)
-            else:
-                output[name] = result
+                result = self._try_cast(result, obj)
+
+            output[idx] = result
+            names.append(name)
 
         if len(output) == 0:
             raise DataError("No numeric types to aggregate")
 
-        return self._wrap_transformed_output(output, names)
+        columns = Index(names)
+        return self._wrap_transformed_output(output, columns)
 
-    def _wrap_aggregated_output(self, output, names=None):
+    def _wrap_aggregated_output(self, output: Mapping[int, np.ndarray], columns: Index):
         raise AbstractMethodError(self)
 
-    def _wrap_transformed_output(self, output, names=None):
+    def _wrap_transformed_output(
+        self, output: Mapping[int, np.ndarray], columns: Index
+    ):
         raise AbstractMethodError(self)
 
     def _wrap_applied_output(self, keys, values, not_indexed_same=False):
         raise AbstractMethodError(self)
 
     def _cython_agg_general(self, how, alt=None, numeric_only=True, min_count=-1):
-        output = {}
+        output = collections.OrderedDict()
+        names = []  # type: List[Hashable]
+
+        # Ideally we would be able to enumerate self._iterate_slices and use
+        # the index from enumeration as the key of output, but ohlc in particular
+        # returns a (n x 4) array. Output requires 1D ndarrays as values, so we
+        # need to slice that up into 1D arrays
+        idx = 0
         for name, obj in self._iterate_slices():
             is_numeric = is_numeric_dtype(obj.dtype)
             if numeric_only and not is_numeric:
                 continue
 
-            result, names = self.grouper.aggregate(obj.values, how, min_count=min_count)
-            output[name] = self._try_cast(result, obj)
+            result, agg_names = self.grouper.aggregate(
+                obj.values, how, min_count=min_count
+            )
+
+            if agg_names:
+                assert len(agg_names) == result.shape[1]
```

Comment:
is this the ohlc case?  if so, comment to that effect?

Reply:
Sure np. This is also noted in the doctoring for `grouper.aggregate`; worth reconsidering (as you've already mentioned) if its really worth changing the signatures of all of these functions just for ohlc